docs/blog/2019-01-04-using-react-context-api-with-gatsby/index.md

```diff
@@ -0,0 +1,273 @@
+
+    const defaultState = {
+      dark: false,
+      toggleDark: () => {},
+    }
+
+    const ThemeContext = React.createContext(defaultState)
+
+    export default ThemeContext
+```
+
+`React.createContext` is a new function in React 16.3 and allows you to create a Context object. It accepts a default state, the value which will be used when a component does not have a matching Provider above it in the tree.  The function returns an object with Provider and Consumer properties which we will be using later.
+
+## Modifying the Gatsby Browser file
+
+Next, write the following code within the `gatsby-browser.js` file, which is in the root folder in a Gatsby project:
+```jsx:title=gatsby-browser.js
+    import React from 'react'
+
+    import ThemeContext from './src/context/ThemeContext'
+
+    class ThemeProvider extends React.Component {
+      state = {
+        dark: false,
+      }
+
+      toggleDark = () => {
+        let dark = !this.state.dark
+        localStorage.setItem('dark', JSON.stringify(dark))
+        this.setState({ dark })
+      }
+
+      componentDidMount() {
+        this.setState({ dark: JSON.parse(localStorage.getItem('dark')) })
+      }
+
+      render() {
+        const { children } = this.props
+        const { dark } = this.state
+        return (
+          <ThemeContext.Provider
+            value={{
+              dark,
+              toggleDark: this.toggleDark,
+            }}
+          >
+            {children}
+          </ThemeContext.Provider>
+        )
+      }
+    }
+// highlight-start
+    export const wrapRootElement = ({ element }) => (
+      <ThemeProvider>{element}</ThemeProvider>
+    )
+// highlight-end
+```
+Create the `ThemeProvider` component which wraps its' children with `TemeContext.Provider`. This component then wraps the root element and is exported as `wrapRootElement`.
+
+The `toggleDark` function gets the current `state.dark` value and switches the value to the opposite. It then stores the new value in `localStorage` before setting it back to state using the `setState` function, so that it persists over page refreshes. The dark value from `state` and the `togglDark` function are passed to the Provider.
+
+## Editing the Layout file
+
+The default `layout.js` uses a `<staticQuery>` and renderProp to render the layout, which is wrapped by a Fragment `<>`. Modify it to look like this:
+```jsx:title=src/components/layout.js
+    import React from 'react'
+    import PropTypes from 'prop-types'
+    import { StaticQuery, graphql } from 'gatsby'
+
+    import ThemeContext from '../context/ThemeContext'
+    import Header from './header'
+    import './layout.css'
+
+    const Layout = ({ children }) => (
+      <StaticQuery
+        query={graphql`
+          query SiteTitleQuery {
+            site {
+              siteMetadata {
+                title
+              }
+            }
+          }
+        `}
+        render={data => (
+          // highlight-start
```

Comment:
This isn't valid JSX, could you do something like:

```jsx
{/* highlight-start */}
<asdfasdf>
{/* highlight-end */}
```

Reply:
Done